### DISCUSSION TITLE
undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState

### DISCUSSION THREAD

**Discussion by rsehgalaka** [abseil/abseil-cpp#1719] [General]
Hello,

I am trying to upgrade GRPC in my project from 1.48.0 to 1.53.0. I have made the required code changes and don't see any complication error but been stuck with this linking error. I think the library that has this "undefined symbol" is `libabsl_strings` which I am linking against as you can see below. Can someone please help me figure out how to resolve this linking issue? I am building on CentOS/gcc 7.3.1 if that helps. Thanks.

Building for CentOS..
`
/opt/rh/devtoolset-7/root/bin/g++ -std=c++17 -Wno-sign-compare -Wno-unused-parameter -Wall -Wextra -fconcepts -DGRPC_1_53_0 -DNOMINMAX -maes -msse4.1 -g -O2 -DGTEST -DTHE_OS='"centos"' -o test_bin libkv.a client_test.o common_definitions_test.o config_test.o backoff_calculator_test.o channel_test.o keyspace_test.o kv_error_test.o logger_test.o main_test.o option_parser_test.o test_data.o token_integration_test.o utils_integration_test.o watch_integration_test.o yaml_parser_test.o -L./grpc_1530 -L/workspace/third_party/centos/grpc_1530/lib -L/workspace/third_party/centos/lib -L/workspace/third_party/centos/grpc_1530/lib64 -L/workspace/third_party/centos/lib64 -L./ -Wl,--start-group -L./ -lkv -pthread -lprotobuf -lgrpc++ -lgrpc -laddress_sorting -lre2 -lupb -lcares -lz -lgpr -labsl_statusor -lssl -lcrypto -labsl_raw_hash_set -labsl_hash -labsl_hashtablez_sampler -labsl_city -labsl_low_level_hash -labsl_random_distributions -labsl_random_seed_sequences -labsl_random_internal_pool_urbg -labsl_random_internal_randen -labsl_random_seed_gen_exception -labsl_random_internal_randen_hwaes -labsl_random_internal_randen_slow -labsl_random_internal_randen_hwaes_impl -labsl_random_internal_platform -labsl_random_internal_seed_material -labsl_status -labsl_cord -labsl_strerror -labsl_cordz_info -labsl_cord_internal -labsl_cordz_functions -labsl_cordz_handle -labsl_exponential_biased -labsl_synchronization -labsl_bad_optional_access -labsl_str_format_internal -labsl_graphcycles_internal -labsl_stacktrace -labsl_symbolize -labsl_time -labsl_debugging_internal -labsl_demangle_internal -labsl_malloc_internal -labsl_civil_time -labsl_strings -labsl_time_zone -labsl_strings_internal -labsl_int128 -labsl_throw_delegate -lrt -labsl_base -labsl_spinlock_wait -labsl_bad_variant_access -labsl_raw_logging_internal -labsl_log_severity -ldl -lstdc++fs -lyaml-cpp -lkvproto -lspdlog -lgtest -lgtest_main -lgmock -lpthread -Wl,--end-group 
`
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord.a(cord.cc.o): In function `absl::lts_20230125::Cord::SetCrcCordState(absl::lts_20230125::crc_internal::CrcCordState)':
cord.cc:(.text+0x2ac2): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState(absl::lts_20230125::crc_internal::CrcCordState&&)'
cord.cc:(.text+0x2ae3): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2b46): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState(absl::lts_20230125::crc_internal::CrcCordState&&)'
cord.cc:(.text+0x2b69): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2bc7): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState(absl::lts_20230125::crc_internal::CrcCordState&&)'
cord.cc:(.text+0x2bf5): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2c29): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2c43): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2c5d): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord.a(cord.cc.o): In function `absl::lts_20230125::Cord::SetExpectedChecksum(unsigned int)':
cord.cc:(.text+0x2c9c): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState()'
cord.cc:(.text+0x2d0b): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState(absl::lts_20230125::crc_internal::CrcCordState&&)'
cord.cc:(.text+0x2d2a): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2d36): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2d47): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
cord.cc:(.text+0x2d58): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord.a(cord.cc.o): In function `absl::lts_20230125::Cord::ExpectedChecksum() const':
cord.cc:(.text+0x2e5c): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::Checksum() const'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord.a(cord.cc.o): In function `absl::lts_20230125::DumpNode(absl::lts_20230125::cord_internal::CordRep*, bool, std::ostream*, int)':
cord.cc:(.text+0x42e1): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::Checksum() const'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord.a(cord.cc.o): In function `absl::lts_20230125::cord_internal::CordRepCrc::~CordRepCrc()':
cord.cc:(.text._ZN4absl12lts_2023012513cord_internal10CordRepCrcD2Ev[_ZN4absl12lts_2023012513cord_internal10CordRepCrcD5Ev]+0x18): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::~CrcCordState()'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord_internal.a(cord_rep_crc.cc.o): In function `absl::lts_20230125::cord_internal::CordRepCrc::New(absl::lts_20230125::cord_internal::CordRep*, absl::lts_20230125::crc_internal::CrcCordState)':
cord_rep_crc.cc:(.text+0x78): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::operator=(absl::lts_20230125::crc_internal::CrcCordState&&)'
cord_rep_crc.cc:(.text+0x128): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::operator=(absl::lts_20230125::crc_internal::CrcCordState&&)'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_cord_internal.a(cord_rep_crc.cc.o): In function `absl::lts_20230125::cord_internal::CordRepCrc::CordRepCrc()':
cord_rep_crc.cc:(.text._ZN4absl12lts_2023012513cord_internal10CordRepCrcC2Ev[_ZN4absl12lts_2023012513cord_internal10CordRepCrcC5Ev]+0x24): undefined reference to `absl::lts_20230125::crc_internal::CrcCordState::CrcCordState()'
collect2: error: ld returned 1 exit status
make: *** [Makefile:46: test_bin] Error 1

**Comment by derekmauro** [comment]
Try adding `-labsl_crc_cord_state` since that is what defines `CrcCordState`:
https://github.com/abseil/abseil-cpp/blob/eb852207758a773965301d0ae717e4235fc5301a/absl/crc/CMakeLists.txt#L149-L163

**Reply by rsehgalaka** [reply]
Thanks!! That moved me a little bit forward. Including `-labsl_crc_cord_state` now gives me. Is there a way I can figure this out on my own, I would love to learn that.

`
/workspace/third_party/centos/grpc_1530/lib64/libabsl_crc_cord_state.a(crc_cord_state.cc.o): In function `absl::lts_20230125::crc_internal::CrcCordState::Checksum() const':
crc_cord_state.cc:(.text+0x3a7): undefined reference to `absl::lts_20230125::RemoveCrc32cPrefix(absl::lts_20230125::crc32c_t, absl::lts_20230125::crc32c_t, unsigned long)'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_crc_cord_state.a(crc_cord_state.cc.o): In function `absl::lts_20230125::crc_internal::CrcCordState::NormalizedPrefixCrcAtNthChunk(unsigned long) const':
crc_cord_state.cc:(.text+0x4a9): undefined reference to `absl::lts_20230125::RemoveCrc32cPrefix(absl::lts_20230125::crc32c_t, absl::lts_20230125::crc32c_t, unsigned long)'
/workspace/third_party/centos/grpc_1530/lib64/libabsl_crc_cord_state.a(crc_cord_state.cc.o): In function `absl::lts_20230125::crc_internal::CrcCordState::Normalize()':
crc_cord_state.cc:(.text+0x5ca): undefined reference to `absl::lts_20230125::RemoveCrc32cPrefix(absl::lts_20230125::crc32c_t, absl::lts_20230125::crc32c_t, unsigned long)'
collect2: error: ld returned 1 exit status
make: *** [Makefile:46: test_bin] Error 1
`

**Reply by derekmauro** [reply]
Use the same logic I used above to figure out what library you need.

What `.cc` file is the missing method (`RemoveCrc32cPrefix`) defined in? Which library is that file part of?

Or use a supported build system, or `pkg-config`, so you don't have to figure this out yourself.

**Reply by rsehgalaka** [reply]
I did use `pkg-config --libs protobuf grpc++ grpc` but apparently it doesn't have all the libraries.
I think I don't know how to get the answer to `Which library is that file part of?` wouldn't I have to go through all the Makefiles (or cmake) in that case? Is it going to be `-lcrc32c` https://github.dev/abseil/abseil-cpp/blob/master/absl/crc/crc32c_test.cc#L182

**Reply by rsehgalaka** [reply]
Thanks @derekmauro I have resolved all errors except for this one
`./grpc_1530/libkvproto.a(kv.grpc.pb.o): In function absl::lts_20210324::MutexLock::MutexLock(absl::lts_20210324::Mutex*)':
kv.grpc.pb.cc:(.text._ZN4absl12lts_202103249MutexLockC2EPNS0_5MutexE[_ZN4absl12lts_202103249MutexLockC5EPNS0_5MutexE]+0x26): undefined reference to `absl::lts_20210324::Mutex::Lock()'
./grpc_1530/libkvproto.a(kv.grpc.pb.o): In function `absl::lts_20210324::MutexLock::~MutexLock()':
`
I am including [labsl_synchronization](https://github.com/abseil/abseil-cpp/blob/eb852207758a773965301d0ae717e4235fc5301a/absl/synchronization/CMakeLists.txt#L71-L72) but it refuses to go away. Can you please help me figure out this one? [stackoverflow](https://stackoverflow.com/questions/76782135/error-when-linking-an-application-against-a-library-using-grpc) didn't help either

**Reply by rsehgalaka** [reply]
Looks like it really isn't defined in libabsl_synchronization
`nm /workspace/third_party/centos/grpc_1530/lib64/libabsl_synchronization.a | grep  "absl.*lts_
20210324.*MutexLock.*MutexLock"
`

**Reply by derekmauro** [reply]
You are likely running into what I described here: https://github.com/abseil/abseil-cpp/issues/1624#issuecomment-1967655080

You either need to build **all** of your code with `-DNDEBUG` or none of it.

**Reply by rsehgalaka** [reply]
Yeah, we don't use `-DNDEBUG` at all. Got any other pointers? The same build was working with GRPC 1.48.0.

**Reply by derekmauro** [reply]
Did you install Abseil with CMake using the Release configuration? If so, you used `-DNDEBUG` without knowing it.  `-DNDEBUG` is standard for Release builds.

This is why I'm telling you to use a supported build system. https://abseil.io/resources/swe-book/html/ch18.html#what_happens_without_a_build_systemques

**Reply by rsehgalaka** [reply]
I didn't specify release configuration. Not sure what it defaults to when I follow instructions [here](https://grpc.io/docs/languages/cpp/quickstart/)
```sh
cmake -DgRPC_INSTALL=ON \
      -DgRPC_BUILD_TESTS=OFF \
      -DCMAKE_INSTALL_PREFIX=$MY_INSTALL_DIR \
      ../..
```

**Reply by rsehgalaka** [reply]
I found something very interesting. All my changes work fine on Ubuntu 18 (gcc 9.40) whereas on CentOS7 (gcc 7.3.1) they don't! Thanks for your help.
